src/platform/posix/posix_pollq_kqueue.c

```diff
@@ -0,0 +1,309 @@
+	nni_posix_pollq_node *wait;   // cancel waiting on this
+	nni_posix_pollq_node *active; // active node (in callback)
+};
+
+int
+nni_posix_pollq_add(nni_posix_pollq *pq, nni_posix_pollq_node *node)
+{
+	if (pq == NULL) {
+		return (NNG_EINVAL);
+	}
+	if (pq->close) {
+		// This shouldn't happen!
+		nni_mtx_unlock(&pq->mtx);
+		return (NNG_ECLOSED);
+	}
+
+	nni_mtx_lock(&pq->mtx);
+	node->pq = pq;
+	nni_mtx_unlock(&pq->mtx);
+
+	return (0);
+}
+
+void
+nni_posix_pollq_remove(nni_posix_pollq_node *node)
+{
+	nni_posix_pollq *pq = node->pq;
+	if (pq == NULL) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+	while (pq->active == node) {
+		pq->wait = node;
+		nni_cv_wait(&pq->cv);
+	}
+
+	if (pq->close) {
+		nni_cv_wake(&pq->cv);
+	}
+	node->pq = NULL;
+	nni_mtx_unlock(&pq->mtx);
+}
+
+void
+nni_posix_pollq_arm(nni_posix_pollq_node *node, int events)
+{
+	nni_posix_pollq *pq = node->pq;
+	struct kevent    kevents[2];
+	int              nevents = 0;
+
+	if (pq == NULL || events == 0) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+
+	if (!(node->events & POLLIN) && (events & POLLIN)) {
+		EV_SET(&kevents[nevents++], node->fd, EVFILT_READ, EV_ADD, 0,
+		    0, (kevent_udata_t) node);
+	}
+
+	if (!(node->events & POLLOUT) && (events & POLLOUT)) {
+		EV_SET(&kevents[nevents++], node->fd, EVFILT_WRITE, EV_ADD, 0,
+		    0, (kevent_udata_t) node);
+	}
+
+	if (nevents > 0) {
+		int rv;
+		rv = kevent(pq->kq, kevents, nevents, NULL, 0, NULL);
+		if (rv < 0) {
+			// report error?
+			fprintf(stderr, "kevent err: %s (%d)", strerror(errno),
+			    errno);
+		} else {
+			node->events |= events;
+		}
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+void
+nni_posix_pollq_disarm(nni_posix_pollq_node *node, int events)
+{
+	struct kevent kevents[2];
+	int           nevents = 0;
+
+	nni_posix_pollq *pq = node->pq;
+	if (pq == NULL) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+
+	if ((node->events & POLLIN) && (events & POLLIN)) {
+		EV_SET(&kevents[nevents++], node->fd, EVFILT_READ, EV_DELETE,
+		    0, 0, (kevent_udata_t) node);
+	}
+
+	if ((node->events & POLLOUT) && (events & POLLOUT)) {
+		EV_SET(&kevents[nevents++], node->fd, EVFILT_WRITE, EV_DELETE,
+		    0, 0, (kevent_udata_t) node);
+	}
+
+	if (nevents > 0) {
+		int rv;
+		rv = kevent(pq->kq, kevents, nevents, NULL, 0, NULL);
+		if (rv < 0) {
+			// report error?
+			fprintf(stderr, "disarm err: %s (%d)\n",
+			    strerror(errno), errno);
+		} else {
+			node->events &= ~events;
+		}
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+static void
+nni_bsd_kqueue_thr(void *arg)
+{
+	nni_posix_pollq *pq        = arg;
+	const int        maxevents = 64;
+	struct kevent    kevents[maxevents];
+
+	nni_mtx_lock(&pq->mtx);
+
+	while (!pq->close) {
+		int i;
+		int nevents;
+
+		// block indefinitely, timers are handled separately
+		nni_mtx_unlock(&pq->mtx);
+		nevents = kevent(pq->kq, NULL, 0, kevents, maxevents, NULL);
+		nni_mtx_lock(&pq->mtx);
+
+		if (nevents < 0) {
+			// report error?
```

Comment:
This is an unfortunate place to have this kind of error... I'm not sure how best to deal with it....  can this really fail?  My read of the man page makes me thing any error other than EINTR here is impossible.

Reply:
The more I think about this the more I think we need to have a new error flag or state on the node, and let the poller set this.  If we have an error here, we would trigger that error on *every* item in the pollq.

This may help other poller implementations that have a need to report some other kind of weird failure scenario.